source/fuzz/transformation_permute_function_parameters.cpp

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <unordered_set>
+#include <vector>
+
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/transformation_permute_function_parameters.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationPermuteFunctionParameters::TransformationPermuteFunctionParameters(
+    const spvtools::fuzz::protobufs::TransformationPermuteFunctionParameters& message)
+    : message_(message) {}
+
+TransformationPermuteFunctionParameters::TransformationPermuteFunctionParameters(
+    uint32_t function_id,
+    uint32_t fresh_type_id,
+    const std::vector<uint32_t>& permutation) {
+  message_.set_function_id(function_id);
+  message_.set_fresh_type_id(fresh_type_id);
+
+  for (auto index : permutation) {
+    message_.add_permutation(index);
+  }
+}
+
+bool TransformationPermuteFunctionParameters::IsApplicable(
+    opt::IRContext* context, const FactManager& /*unused*/) const {
+  // Check that function exists
+  auto* function = fuzzerutil::FindFunction(context, message_.function_id());
+  if (!function ||
+      function->DefInst().opcode() != SpvOpFunction ||
+      fuzzerutil::FunctionIsEntryPoint(context, function->result_id())) {
+    return false;
+  }
+
+  if (!fuzzerutil::IsFreshId(context, message_.fresh_type_id())) {
+    return false;
+  }
+
+  // Check that permutation has valid indices
+  auto* function_type = fuzzerutil::GetFunctionType(context, function);
+  assert(function_type && "Function type is null");
+
+  const auto& permutation = message_.permutation();
+
+  // Don't take return type into account
+  uint32_t arg_size = function_type->NumInOperands() - 1;
+
+  if (static_cast<uint32_t>(permutation.size()) != arg_size) {
+    return false;
+  }
+
+  std::unordered_set<uint32_t> set;
+  for (auto index : permutation) {
+    // Don't compare an unsigned integer with 0
```

Comment:
I don't get this comment in context.

Reply:
I'm just checking that all indices are valid. Thus, the full check is `index < 0 || index >= arg_size`.
But, since `index` is `uint32_t`, compiler complains that we needlessly compare unsigned integer with 0.

Reply:
Oh I see; in that case I would change it to say that the value is non-negative by definition, or just remove the comment, as you prefer.

Reply:
Fixed in eb526e4